### PR TITLE
People/davidgamba/rmd 20676 refactor required and optional task arguments 2

### PR DESCRIPTION
**Pull request by DavidGamba**
@tonytvo please review.

### CODE REVIEW

**Comment by tonytvo** [comment]
here's the list of java command line parser: https://stackoverflow.com/questions/367706/how-to-parse-command-line-arguments-in-java

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeTask.groovy [hunk 1]
```diff
@@ -44,50 +44,45 @@ abstract class VeracodeTask extends DefaultTask {
...
-    final boolean haveRequiredArguments() {
-        boolean haveRequiredArguments = true
-        requiredArguments.each() {
-            if (!isArgumentOptional(it)) {
-                haveRequiredArguments &= getProject().hasProperty(it)
-            }
+    protected boolean hasRequiredArguments() {
+        boolean hasRequiredArguments = true
+        requiredArguments.each() { arg ->
+            hasRequiredArguments &= getProject().hasProperty(arg)
```
Comment (tonytvo) [inline src/main/groovy/com/calgaryscientific/gradle/VeracodeTask.groovy:76]: optional, you can try to use groovy idiom here: requiredArguments.any { arg -> !getProject).hasProperty(arg) }

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeTask.groovy [hunk 2]
```diff
@@ -44,50 +44,45 @@ abstract class VeracodeTask extends DefaultTask {
...
-            } else {
-                String originalArgument = it.substring(0, it.length() - OPTIONAL.length())
-                sb.append(" [-P${originalArgument}=${validArguments.get(originalArgument)}]")
-            }
+    protected static String correctUsage(String taskName,
+                                         List<String> requiredArguments,
+                                         List<String> optionalArguments) {
+        StringBuilder sb = new StringBuilder("Missing required arguments: gradle ${taskName}")
+        requiredArguments.each() { arg ->
+            sb.append(" -P${arg}=${validArguments.get(arg)}")
```
Comment (tonytvo) [inline src/main/groovy/com/calgaryscientific/gradle/VeracodeTask.groovy:65]: I would do something with groovy idiom: requiredArguments.collect { arg -> -P${arg}=${validArguments.get(arg)} }.join(' ')

**Review by tonytvo — APPROVED** [review]

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeTask.groovy (hunk 2 above)
Comment (DavidGamba) [reply src/main/groovy/com/calgaryscientific/gradle/VeracodeTask.groovy:65]: It doesn't gain much readability in this case I don't think:

````groovy
        requiredArguments.each() { arg ->
            sb.append(" -P${arg}=${validArguments.get(arg)}")
        }
````
vs
````groovy
        sb.append(" " + requiredArguments.collect { arg ->
            "-P${arg}=${validArguments.get(arg)}"}.join(" ")
        )
````
or
````groovy
        sb.append(" ")
        sb.append(requiredArguments.collect { arg ->
            "-P${arg}=${validArguments.get(arg)}"}.join(" ")
        )
````

**Pull request merged**